### PR TITLE
DRDT-106: Ad Stack: Salesforce DMP

### PR DESCRIPTION
**Pull request by akhan3051**
[DRDT-106](https://readersdigest.atlassian.net/browse/DRDT-106)

### CODE REVIEW

**Comment by fariasf** [comment]
Created new staging environment for pr-78. Here are some useful links:

[WordPress Admin](https://pr-78-drdt.pantheonsite.io/wp/wp-admin)

[Homepage](https://pr-78-drdt.pantheonsite.io/)
[Article](https://pr-78-drdt.pantheonsite.io/jobsite-tips/abandoned-cable-removal/)
[Listicle](https://pr-78-drdt.pantheonsite.io/jobsite-tips/exterior/12-scaffolding-safety-tips-and-handy-hints/)
[Page](https://pr-78-drdt.pantheonsite.io/about-construction-pro-tips/)
[Hub](https://pr-78-drdt.pantheonsite.io/jobsite-tips/)

**Comment by fariasf** [comment]
There are visual regressions! 

View the [full visual regression test report](https://circleci.com/gh/fariasf/drdt-test/2615/artifacts/0/artifacts/backstop_data/html_report/index.html)

**Comment by fariasf** [comment]
⚠ Frontend performance tests failed!

Full Lighthouse performance results:

| Page/Branch | pr-78 | master | Result |
| --- | ---: | ---: | :---: |
| Homepage | [92](https://circleci.com/gh/fariasf/drdt-test/2616/artifacts/0/artifacts/lighthouse_results/pr-78/homepage/lighthouse.report.html) | [96](https://circleci.com/gh/fariasf/drdt-test/2616/artifacts/0/artifacts/lighthouse_results/master/homepage/lighthouse.report.html) | ✔ |
| Article | [89](https://circleci.com/gh/fariasf/drdt-test/2616/artifacts/0/artifacts/lighthouse_results/pr-78/article/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2616/artifacts/0/artifacts/lighthouse_results/master/article/lighthouse.report.html) | ✘ |
| Listicle | [97](https://circleci.com/gh/fariasf/drdt-test/2616/artifacts/0/artifacts/lighthouse_results/pr-78/listicle/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2616/artifacts/0/artifacts/lighthouse_results/master/listicle/lighthouse.report.html) | ✔ |
| Page | [98](https://circleci.com/gh/fariasf/drdt-test/2616/artifacts/0/artifacts/lighthouse_results/pr-78/page/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2616/artifacts/0/artifacts/lighthouse_results/master/page/lighthouse.report.html) | ✔ |
| Hub | [97](https://circleci.com/gh/fariasf/drdt-test/2616/artifacts/0/artifacts/lighthouse_results/pr-78/hub/lighthouse.report.html) | [93](https://circleci.com/gh/fariasf/drdt-test/2616/artifacts/0/artifacts/lighthouse_results/master/hub/lighthouse.report.html) | ✔ |


(**Note: Tests are against the _NOADS_ version of the pages!**)

**Review by fariasf — CHANGES_REQUESTED** [review]
This PR includes a lot of plugins that shouldn't be here. @akhan3051 are you starting branches from the latest code from master?

File: web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+	/**
+	 * Add the salesforce tags
+	 */
+	public static function render_salesforce_dmp_tags() {
+
+		$control_tag_script = PHP_EOL . "window.Krux||((Krux=function(){Krux.q.push(arguments)}).q=[]);
+  (function(){
+    var k=document.createElement('script');k.type='text/javascript';k.async=true;
+    k.src=(location.protocol==='https:'?'https:':'http:')+'//cdn.krxd.net/controltag/s9xpab5u5.js';
```
Comment (fariasf) [inline web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php:86]: `s9xpab5u5` also appears as part as `CONTROL_SCRIPT_TAG`. Should this be moved to config, to support other sites?

File: web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+		if ( ! is_single() ) {
+			return array();
+		}
+
+		$top_level_categories = '';
+		$sub_categories       = '';
+		$categories           = array();
+		$keywords             = '';
+
+		if ( class_exists( 'Ad_Stack' ) ) {
```
Comment (fariasf) [inline web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php:148]: Why does this depend on the `Ad_Stack`?

File: web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+		self::$krux_data_layer['page']['author']        = self::get_post_authors();
+		self::$krux_data_layer['page']['gs_categories'] = json_decode( get_post_meta( $content_id, 'gs_channels', true ) );
+
+		$tax_array = self::get_post_taxonomy();
+		if ( $tax_array && ( class_exists( 'Ad_Stack' ) ) ) {
+			foreach ( $tax_array as $tax => $value ) {
+				self::$krux_data_layer['page'][ $tax ] = self::format_data( $value );
+			}
+		}
+		return self::$krux_data_layer;
```
Comment (fariasf) [inline web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php:179]: This function does too much and knows too much about its environment (e.g. the `gs_channels` meta field, getting info from taxonomies, tags, etc). Consider breaking it down in smaller functions, and collect the results from each of them via filters.

File: web/wp-content/plugins/tmbi-salesforce-dmp/inc/author-detail.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
+<?php
```
Comment (fariasf) [inline web/wp-content/plugins/tmbi-salesforce-dmp/inc/author-detail.php:1]: Functions in this file have generic names without a prefix or namespace. Consider namespacing them, or prefix their names with the plugin name.

File: web/wp-content/plugins/tmbi-salesforce-dmp/js/krux-click-traking.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
+jQuery(document).ready(function($) {
```
Comment (fariasf) [inline web/wp-content/plugins/tmbi-salesforce-dmp/js/krux-click-traking.js:1]: This file should be part of the theme, as the markup is decided from the theme layer. E.g. an element with class `gform_wrapper` containing a child element with class `newsletter-signup` - If we keep it here, when we alter the theme's CSS or HTML structure we'll need to rewrite this part of the plugin.

**Comment by fariasf** [comment]
Updated staging environment for pr-78. Here are some useful links:

[WordPress Admin](https://pr-78-drdt.pantheonsite.io/wp/wp-admin)

[Homepage](https://pr-78-drdt.pantheonsite.io/)
[Article](https://pr-78-drdt.pantheonsite.io/jobsite-tips/abandoned-cable-removal/)
[Listicle](https://pr-78-drdt.pantheonsite.io/jobsite-tips/exterior/12-scaffolding-safety-tips-and-handy-hints/)
[Page](https://pr-78-drdt.pantheonsite.io/about-construction-pro-tips/)
[Hub](https://pr-78-drdt.pantheonsite.io/jobsite-tips/)

**Comment by fariasf** [comment]
There are visual regressions! 

View the [full visual regression test report](https://circleci.com/gh/fariasf/drdt-test/2672/artifacts/0/artifacts/backstop_data/html_report/index.html)

**Comment by fariasf** [comment]
Congrats! Frontend performance tests passed!

Full Lighthouse performance results:

| Page/Branch | pr-78 | master | Result |
| --- | ---: | ---: | :---: |
| Homepage | [95](https://circleci.com/gh/fariasf/drdt-test/2671/artifacts/0/artifacts/lighthouse_results/pr-78/homepage/lighthouse.report.html) | [96](https://circleci.com/gh/fariasf/drdt-test/2671/artifacts/0/artifacts/lighthouse_results/master/homepage/lighthouse.report.html) | ✔ |
| Article | [98](https://circleci.com/gh/fariasf/drdt-test/2671/artifacts/0/artifacts/lighthouse_results/pr-78/article/lighthouse.report.html) | [96](https://circleci.com/gh/fariasf/drdt-test/2671/artifacts/0/artifacts/lighthouse_results/master/article/lighthouse.report.html) | ✔ |
| Listicle | [97](https://circleci.com/gh/fariasf/drdt-test/2671/artifacts/0/artifacts/lighthouse_results/pr-78/listicle/lighthouse.report.html) | [97](https://circleci.com/gh/fariasf/drdt-test/2671/artifacts/0/artifacts/lighthouse_results/master/listicle/lighthouse.report.html) | ✔ |
| Page | [98](https://circleci.com/gh/fariasf/drdt-test/2671/artifacts/0/artifacts/lighthouse_results/pr-78/page/lighthouse.report.html) | [97](https://circleci.com/gh/fariasf/drdt-test/2671/artifacts/0/artifacts/lighthouse_results/master/page/lighthouse.report.html) | ✔ |
| Hub | [97](https://circleci.com/gh/fariasf/drdt-test/2671/artifacts/0/artifacts/lighthouse_results/pr-78/hub/lighthouse.report.html) | [94](https://circleci.com/gh/fariasf/drdt-test/2671/artifacts/0/artifacts/lighthouse_results/master/hub/lighthouse.report.html) | ✔ |


(**Note: Tests are against the _NOADS_ version of the pages!**)

File: web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php (hunk 1 above)
Comment (akhan3051) [reply web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php:86]: @fariasf as this is same for all our brands except for RDCA. I thought of keeping it here.

File: web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+				'data-id'        => [],
+				'data-timing'    => [],
+				'data-version'   => [],
+				'type'           => [],
+				'data-namespace' => [],
+			],
+		];
+		$interchage_tag = sprintf( PHP_EOL . self::INTERCHANGE_SCRIPT_TAG, $interchage_tag_script );
+		$variant        = get_query_var( 'variant' );
+		if ( 'noads' !== $variant ) {
```
Comment (fariasf) [inline web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php:127]: After this is merged, we can integrate it with the Ad Blocker plugin, so we don't have to maintain that logic here.

**Review by fariasf — CHANGES_REQUESTED** [review]
@akhan3051 it seems this Krux plugin depends on the `Ad_Stack` class (there are several checks for its existence). However, there's no dependency at a code level (class doesn't seem to be invoked). If the dependency is because of business requirements (this should only run if the Ad Stack is running) please make that clear. Otherwise I don't see the point on keeping an artificial dependency (if we rename the `Ad_Stack` class for whatever reason, this code will stop working as intended).

**Comment by fariasf** [comment]
Closing this for now (to save us a multidev environment). Need to address the changes and fix the line endings. @akhan3051 please reopen if you can address any of those issues. I moved the ticket back to In Progress. We can review during standup.

**Pull request closed without merging**